Add hydrological forcing comparison recipe

stefsmeets opened the pull request:
## Description

Hi everyone, this recipe compares between ERA5, ERA-interim, and MSWEP datasets for a certain catchment.

- Closes #issue_number
- Link to documentation: https://esmvaltool--2013.org.readthedocs.build/en/2013/recipes/recipe_hydro_forcing.html 

## Todo:

- [x] https://github.com/ESMValGroup/ESMValCore/pull/986 merged

* * *

## Before you get started

- [ ] [☝ Create an issue](https://github.com/ESMValGroup/ESMValTool/issues) to discuss what you are going to do

## Checklist

It is the responsibility of the author to make sure the PR is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

- [x] [🛠][1] PR has a descriptive title for the [changelog](https://docs.esmvaltool.org/en/latest/community/introduction.html#branches-pull-requests-and-code-review)
- [x] [🛠][1] Code follows the [style guide](https://docs.esmvaltool.org/en/latest/community/introduction.html#code-style)
- [x] [🛠][1] [Documentation](https://docs.esmvaltool.org/en/latest/community/introduction.html#documentation) is available
- [x] [🛠][1] YAML files pass [`pre-commit`](https://docs.esmvaltool.org/en/latest/community/introduction.html#pre-commit) or [`yamllint`](https://docs.esmvaltool.org/en/latest/community/introduction.html#yaml) checks
- [x] [🛠][1] [Circle/CI tests pass](https://docs.esmvaltool.org/en/latest/community/introduction.html#branches-pull-requests-and-code-review)
- [x] [🛠][1] [Codacy code quality checks pass](https://docs.esmvaltool.org/en/latest/community/introduction.html#branches-pull-requests-and-code-review)
- [x] [🛠][1] [Documentation builds successfully](https://docs.esmvaltool.org/en/latest/community/introduction.html#branches-pull-requests-and-code-review) on readthedocs

### New or updated [recipe/diagnostic](https://docs.esmvaltool.org/en/latest/community/diagnostic.html):

- [x] [🛠][1] [Provenance information](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#recording-provenance) has been added
- [x] [🛠][1] New dependencies are added to the [project requirements](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#additional-dependencies)
- [x] [🧪][2] [Documentation](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#documentation) for the recipe/diagnostic clearly describes what it calculates from a scientific point of view
- [ ] [🧪][2] Recipe runs successfully on [`@esmvalbot`](https://docs.esmvaltool.org/en/latest/community/introduction.html#running-tests) or some other machine without modification (MSWEP data not available)
- [x] [🧪][2] Figure(s)/data [look as expected](https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review)
- [x] [🧪][2] Code is [well documented](https://docs.esmvaltool.org/en/latest/community/introduction.html#what-should-be-documented) and scientifically sound

***

To help with the number pull requests:

-  🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValTool/pulls) in this repository

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

stefsmeets commented:
Hi @Peter9192 , I think this PR is ready. Could you do a technical review?

jvegasbsc commented:
I can test it if you want, but I will need a suitable shapefile

stefsmeets commented:
> I can test it if you want, but I will need a suitable shapefile

@jvegasbsc That would be great! I'm using the shapefiles from here: https://github.com/eWaterCycle/recipes_auxiliary_datasets/tree/master/Lorentz_Basin_Shapefiles

You will also need the MSWEP data: https://docs.esmvaltool.org/projects/ESMValCore/en/latest/develop/fixing_data.html?highlight=mswep#mswep
(or just run with ERA5 / ERA-Interim)

jvegasbsc commented:
Thanks. I've just run it for ERA-Interim, as we only have monthly data for ERA5 in jasmin

jvegasbsc commented on esmvaltool/diag_scripts/hydrology/hydro_forcing.py:
```diff
@@ -0,0 +1,203 @@
+    """Load and plot hydro forcing data."""
+    diagnostic, metadata = get_diagnostic_data(cfg)
+
+    if diagnostic == 'sample_year':
```
I really don't like using the name of the diagnostic to define its behaviour. It's obscure and can mislead the users. It also imposes some limitations on the recipe itself, like being unable to do the detailed year plot for more than one year in a single run.

Can you please change it to a user setting and update the doc?

stefsmeets replied:
Thanks, I was not aware of this convention. I had this at first, but it seemed a bit redundant to me to use a setting to pass the name of the diagnostic while it is already available, which is why I looked for an alternative. 

Will change it back 👍

nielsdrost replied:
Thanks @stefsmeets, yes this looks much better.

I noticed you use "entry_point" as the name of the setting. I do not think this is very clear to the user (entry into what?). 

Perhaps change it to "plot_type" as also used in the [Python example](https://github.com/ESMValGroup/ESMValTool/blob/master/esmvaltool/recipes/examples/recipe_python.yml)?

stefsmeets replied:
Sure, I have updated it to `plot_type` 👍

jvegasbsc reviewed (changes requested):
The code itself looks good and clean. Good work, @stefsmeets

stefsmeets commented:
Thanks @jvegasbsc , I have updated the code. I think it is better now 👍

stefsmeets commented:
Hi @jeromaerts , could you do a *scientific review* of this recipe? Especially, could you have a look at the documentation, descriptions, captions, etc., whether they make sense scientifically?

jeromaerts commented:
@stefsmeets, Looks great, also from a scientific point of view. 

Some small notes on the documentation:

- "3. Plot the monthly climate statistics over a certain period"
  Change: "3. Plot the monthly **climatology** statistics over a certain period"

If possible extent the plotting of climatology to also include a daily climatology plot.

- "All hydrological recipes require a shapefile as an input to **produce** forcing data."
    Change, produce to **select**.

- "All recipes are tested with the shapefiles that are used for the eWaterCycle project."
   "All recipes are tested with shapefiles **from HydroSHEDS** that are used for the eWaterCycle project."

Looking at the figures, the dates on the horizontal axis are very close to each other, perhaps slightly reduce the label font-size.

stefsmeets commented:
Hi @jeromaerts , thanks for the comments! I have updated the code and added the new plot. Let me know what you think. If you are happy with the changes, could you [approve the PR](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/approving-a-pull-request-with-required-reviews)?

jeromaerts reviewed (approved):
The docs are updated and consistent with hydrology conventions.

stefsmeets commented:
Hi @jvegasbsc , have you had a change to take another look at the changes I made? If you agree, could you approve this PR so we can get this merged? If you have any other comments, please let me know!

stefsmeets commented:
ping @jvegasbsc 😁

jvegasbsc reviewed (approved):
Looks better now! I agree with @nielsdrost comment about tha nimng of the `entry_point` but that's a very minor concern.

And sorry for the delay, I went straight to holidays after the release

stefsmeets commented:
> Looks better now! I agree with @nielsdrost comment about tha nimng of the `entry_point` but that's a very minor concern.
> 
> And sorry for the delay, I went straight to holidays after the release

No worries, probably well-deserved! 😅 Will fix it next week 👍

nielsdrost commented:
Thanks @stefsmeets !
